### PR TITLE
WIP: Add support for textDocument/codeLens

### PR DESCRIPTION
**Pull request by andreypopp**
This adds implementation of `textDocument/codeLens` request.

The result is rendered through neovim's virtualtext API and thus it's only usable with neovim.

codelens

TODO:
- [ ] Tests (I know!)
- [x] Commands to enable/disable/toggle (also per buffer variants)

### CODE REVIEW

File: autoload/ale.vim [hunk 1]
```diff
@@ -109,6 +109,7 @@ function! s:Lint(buffer, should_lint_file, timer_id) abort
     \   || (a:should_lint_file && filereadable(expand('#' . a:buffer . ':p')))
 
     call ale#engine#RunLinters(a:buffer, l:linters, l:lint_file)
+    call ale#code_lens#Request(a:buffer)
```
Comment (andreypopp) [inline autoload/ale.vim]: I'm not sure this is the right way — what I need is to make requests in the same cases we do requests diagnostics (also after we synced changed to LSP).

File: autoload/ale/code_lens.vim [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+function! ale#code_lens#SetMap(map) abort
+    let s:code_lens_map = a:map
+endfunction
+
+function! ale#code_lens#ClearLSPData() abort
+    let s:code_lens_map = {}
+endfunction
+
+" TODO: maybe check with exists instead?
+let s:supported = exists('*nvim_buf_set_virtual_text')
```
Comment (andreypopp) [inline autoload/ale/code_lens.vim]: You have `has('nvim-0.3.2')` in `virtualtext.vim`, why?

File: autoload/ale/code_lens.vim [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    if !ale#code_lens#IsEnabled(a:buffer)
+        return
+    endif
+
+    " Set a flag so we only make one request.
+    call setbufvar(a:buffer, 'ale_code_lens_request_made', 0)
+
+    for l:linter in ale#linter#Get(getbufvar(a:buffer, '&filetype'))
+        if !empty(l:linter.lsp) && l:linter.lsp isnot# 'tsserver'
+            call s:Request(l:linter, a:buffer)
```
Comment (andreypopp) [inline autoload/ale/code_lens.vim]: This is incorrect now as requests to different LSPs will override each others codelens, I'm not sure though if we need to support the case with multiple LSPs + codelens, thoughts?

File: autoload/ale.vim (hunk 1 above)
Comment (w0rp) [inline autoload/ale.vim]: Check if the lens feature is enabled before calling functions from `code_lens.vim`, and then the code won't be loaded for users who don't turn the feature on.

**Comment by andreypopp** [comment]
Thanks! Addressed first round of feedback.

File: autoload/ale/code_lens.vim (hunk 2 above)
Comment (w0rp) [reply autoload/ale/code_lens.vim]: I can't remember, but I think that was the minimum version requirement due to changes for the functions.

File: autoload/ale/code_lens.vim (hunk 3 above)
Comment (w0rp) [reply autoload/ale/code_lens.vim]: You can either combine the results of multiple servers, or ignore responses from other servers. Note that if you choose to ignore responses, you should make a request to all available servers, but keep responses from only one.

File: autoload/ale.vim [hunk 4]
```diff
@@ -109,7 +109,10 @@ function! s:Lint(buffer, should_lint_file, timer_id) abort
     \   || (a:should_lint_file && filereadable(expand('#' . a:buffer . ':p')))
 
     call ale#engine#RunLinters(a:buffer, l:linters, l:lint_file)
-    call ale#code_lens#Request(a:buffer)
+
+    if get(g:, 'ale_code_lens_enabled', 0)
```
Comment (w0rp) [inline autoload/ale.vim]: You might want to use `ale#Var(a:buffer, 'code_lens_enabled)` here, so the code lens feature can be enabled for a single buffer, while it's disabled globally.

Comment (andreypopp) [reply autoload/ale.vim]: That's useful, thank you!

Do you know how we can guarantee codelens request is sent after didChange notification is sent to server?

Comment (w0rp) [reply autoload/ale.vim]: The easiest way would be to always attempt to send a message about the buffer being changed before every codeLens request, and then the check for `b:changedtick` will prevent the message from being sent more than it needs to be.

**Comment by w0rp** [comment]
I'm closing old pull requests that didn't go anywhere, or I didn't have time to look at. Re-open the pull request or submit a new one if you have something you think can be merged.

**Pull request closed without merging**